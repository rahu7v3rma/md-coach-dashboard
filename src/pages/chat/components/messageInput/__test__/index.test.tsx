import {
    RenderResult,
    act,
    cleanup,
    fireEvent,
    render,
    waitFor
} from '@testing-library/react';
import userEvt from '@testing-library/user-event';
import { StreamChat, User } from 'stream-chat';
import { Channel, Chat, MessageInput, Thread, Window } from 'stream-chat-react';

import CustomInput from '../../messageInput';

const someValues = [{ name: 'teresa teng' }];
const mockUploadNewFiles = jest.fn();
const mockHandleSubmit = jest.fn();

jest.setTimeout(20000);
let tree: RenderResult;

const StreamChatReal = require('stream-chat-react');

const userId = 'weathered-fire-2';
const userName = 'weathered';

const user: User = {
    id: userId,
    name: userName,
    image: `https://getstream.io/random_png/?id=${userId}&name=${userName}`
};

const apiKey = 'dz5f4d5kzrue';
const userToken =
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyX2lkIjoid2VhdGhlcmVkLWZpcmUtMiIsImV4cCI6MTY5NzMxNjYyMH0.VOblsnzaKCGpqz4vPtNzQRJCLY0-SLu_HtWIBdPFTIw';

const chatClient = new StreamChat(apiKey);
chatClient.connectUser(user, userToken);

const channel = chatClient.channel('messaging', 'custom_channel_id', {
    // add as many custom fields as you'd like
    image: 'https://www.drupal.org/files/project-images/react.png',
    name: 'Talk about React',
    members: [userId]
});

const ChatTest = () => (
    <Chat client={chatClient}>
        <Channel channel={channel} Input={CustomInput}>
            <Window>
                <MessageInput />
            </Window>
            <Thread />
        </Channel>
    </Chat>
);

beforeEach(async () => {
    tree = await act(() => render(<ChatTest />));
});

afterEach(cleanup);

describe('test message input', () => {
    test('match Snapshot', async () => {
        await waitFor(
            () =>
                expect(
                    tree.getByPlaceholderText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        expect(tree).toMatchSnapshot();
    });

    test('renders the file upload button with an attach icon', async () => {
        await waitFor(
            () =>
                expect(
                    tree.getByPlaceholderText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        expect(tree.getByAltText('Attach')).toBeInTheDocument();
    });

    test('renders the uploads preview component', async () => {
        await waitFor(
            () =>
                expect(
                    tree.getByLabelText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        const str = JSON.stringify(someValues);
        const blob = new Blob([str]);
        const file = new File([blob], 'values.json', {
            type: 'application/JSON'
        });
        File.prototype.text = jest.fn().mockResolvedValueOnce(str);
        const input = tree.getByLabelText('File input');
        await act(() => userEvt.upload(input, file));
        expect(
            tree.container.querySelector('.rfu-file-previewer')
        ).toBeInTheDocument();
    });

    test('enables multiple uploads', async () => {
        await waitFor(
            () =>
                expect(
                    tree.getByLabelText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        const str = JSON.stringify(someValues);
        const blob = new Blob([str]);
        const file = new File([blob], 'values.json', {
            type: 'application/JSON'
        });
        File.prototype.text = jest.fn().mockResolvedValueOnce(str);
        const input = tree.getByLabelText('File input');
        await act(() => userEvt.upload(input, file));
        await act(() => userEvt.upload(input, file));
        expect(
            tree.container.querySelectorAll('.rfu-file-previewer__file').length
        ).toEqual(2);
    });

    test('calls the uploadNewFiles function when a file is selected', async () => {
        jest.spyOn(StreamChatReal, 'useMessageInputContext').mockImplementation(
            () => ({
                uploadNewFiles: mockUploadNewFiles,
                maxFilesLeft: 10,
                handleSubmit: mockHandleSubmit
            })
        );
        tree = await act(() => render(<ChatTest />));
        await waitFor(
            () =>
                expect(
                    tree.getAllByPlaceholderText('Enter your message')[1]
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        const str = JSON.stringify(someValues);
        const blob = new Blob([str]);
        const file = new File([blob], 'values.json', {
            type: 'application/JSON'
        });
        File.prototype.text = jest.fn().mockResolvedValueOnce(str);
        const input = tree.getAllByLabelText('File input')[1];
        await act(() => userEvt.upload(input, file));
        expect(mockUploadNewFiles).toBeCalled();
    });

    test('disables the file upload button when maxFilesLeft is zero', async () => {
        jest.spyOn(StreamChatReal, 'useMessageInputContext').mockImplementation(
            () => ({
                uploadNewFiles: mockUploadNewFiles,
                maxFilesLeft: 0,
                handleSubmit: mockHandleSubmit
            })
        );
        await waitFor(
            () =>
                expect(
                    tree.getByPlaceholderText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        tree = await act(() => render(<ChatTest />));

        const inputs = tree.getAllByLabelText('File input');
        expect((inputs[1] as HTMLInputElement).disabled).toBeTruthy();
    });

    test('renders the message input with a placeholder', async () => {
        await waitFor(
            () =>
                expect(
                    tree.getByPlaceholderText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
    });

    test('renders the send icon', async () => {
        await waitFor(
            () =>
                expect(
                    tree.getByPlaceholderText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        const div = tree.container.querySelector('.str-chat__textarea');
        expect(
            div?.nextElementSibling?.children[0].hasAttribute('src')
        ).toBeTruthy();
    });

    test('calls the handleSubmit function when the send icon is clicked', async () => {
        jest.spyOn(StreamChatReal, 'useMessageInputContext').mockImplementation(
            () => ({
                uploadNewFiles: mockUploadNewFiles,
                maxFilesLeft: 10,
                handleSubmit: mockHandleSubmit
            })
        );
        await waitFor(
            () =>
                expect(
                    tree.getByPlaceholderText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        const div = tree.container.querySelector('.str-chat__textarea');
        div?.nextElementSibling?.children[0] &&
            fireEvent.click(div?.nextElementSibling?.children[0]);
        expect(mockHandleSubmit).toBeCalled();
    });

    test('renders the send icon', async () => {
        await waitFor(
            () =>
                expect(
                    tree.getByPlaceholderText('Enter your message')
                ).toBeInTheDocument(),
            { timeout: 5000 }
        );
        const div = tree.container.querySelector('.str-chat__textarea');
        expect(
            div?.nextElementSibling?.children[0].hasAttribute('src')
        ).toBeTruthy();
    });
});
