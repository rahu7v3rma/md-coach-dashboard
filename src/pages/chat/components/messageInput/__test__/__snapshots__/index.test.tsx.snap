// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test message input match Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="str-chat str-chat-channel messaging light "
      >
        <div
          class="str-chat__container"
        >
          <div
            class="str-chat__main-panel   "
          >
            <div
              class="
			  rfu-dropzone
			  
			  
			"
              role="button"
              style="position: relative;"
              tabindex="-1"
            >
              <div
                class="rfu-dropzone__notifier"
              >
                <div
                  class="rfu-dropzone__inner"
                >
                  <svg
                    height="41"
                    viewBox="0 0 41 41"
                    width="41"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M40.517 28.002V3.997c0-2.197-1.808-3.992-4.005-3.992H12.507a4.004 4.004 0 0 0-3.992 3.993v24.004a4.004 4.004 0 0 0 3.992 3.993h24.005c2.197 0 4.005-1.795 4.005-3.993zm-22.002-7.997l4.062 5.42 5.937-7.423 7.998 10H12.507l6.008-7.997zM.517 8.003V36c0 2.198 1.795 4.005 3.993 4.005h27.997V36H4.51V8.002H.517z"
                      fill="#000"
                      fill-rule="nonzero"
                    />
                  </svg>
                  <p>
                    Drag your files here to add to your post
                  </p>
                </div>
              </div>
              <div
                class="sc-ispOId dEptGl"
              >
                <div
                  class="rfu-file-upload-button"
                >
                  <label>
                    <input
                      aria-label="File input"
                      class="rfu-file-input"
                      multiple=""
                      type="file"
                    />
                    <img
                      alt="Attach"
                      class="sc-iVCKna gEEioe"
                      src="[object Object]"
                    />
                  </label>
                </div>
                <div
                  class="sc-kcuKUB fHxCqo"
                >
                  <div
                    class="rta  str-chat__textarea"
                  >
                    <textarea
                      aria-label="Enter your message"
                      class="rta__textarea str-chat__textarea__textarea"
                      data-testid="message-input"
                      placeholder="Enter your message"
                      rows="1"
                    />
                  </div>
                  <div
                    class="sc-bgqQcB gckPcd"
                  >
                    <img
                      class="sc-gTRrQi fPcYip"
                      src="[object Object]"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="str-chat str-chat-channel messaging light "
    >
      <div
        class="str-chat__container"
      >
        <div
          class="str-chat__main-panel   "
        >
          <div
            class="
			  rfu-dropzone
			  
			  
			"
            role="button"
            style="position: relative;"
            tabindex="-1"
          >
            <div
              class="rfu-dropzone__notifier"
            >
              <div
                class="rfu-dropzone__inner"
              >
                <svg
                  height="41"
                  viewBox="0 0 41 41"
                  width="41"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M40.517 28.002V3.997c0-2.197-1.808-3.992-4.005-3.992H12.507a4.004 4.004 0 0 0-3.992 3.993v24.004a4.004 4.004 0 0 0 3.992 3.993h24.005c2.197 0 4.005-1.795 4.005-3.993zm-22.002-7.997l4.062 5.42 5.937-7.423 7.998 10H12.507l6.008-7.997zM.517 8.003V36c0 2.198 1.795 4.005 3.993 4.005h27.997V36H4.51V8.002H.517z"
                    fill="#000"
                    fill-rule="nonzero"
                  />
                </svg>
                <p>
                  Drag your files here to add to your post
                </p>
              </div>
            </div>
            <div
              class="sc-ispOId dEptGl"
            >
              <div
                class="rfu-file-upload-button"
              >
                <label>
                  <input
                    aria-label="File input"
                    class="rfu-file-input"
                    multiple=""
                    type="file"
                  />
                  <img
                    alt="Attach"
                    class="sc-iVCKna gEEioe"
                    src="[object Object]"
                  />
                </label>
              </div>
              <div
                class="sc-kcuKUB fHxCqo"
              >
                <div
                  class="rta  str-chat__textarea"
                >
                  <textarea
                    aria-label="Enter your message"
                    class="rta__textarea str-chat__textarea__textarea"
                    data-testid="message-input"
                    placeholder="Enter your message"
                    rows="1"
                  />
                </div>
                <div
                  class="sc-bgqQcB gckPcd"
                >
                  <img
                    class="sc-gTRrQi fPcYip"
                    src="[object Object]"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
